lib/query.js

```diff
@@ -167,6 +192,32 @@ function topLevelFuncDecls(parsed) {
   return FindToplevelFuncDeclVisitor.run(parsed);
 }
 
+function resolveReference(ref, scopePath) {
+  if (scopePath.length == 0) return [null, null];
+  const [scope, ...outer] = scopePath;
+  const decls = declarationsWithIdsOfScope(scope, true);
+  const decl = decls.find(([_, id]) => id.name == ref);
+  return decl || resolveReference(ref, outer);
+}
+
+function resolveReferences(scope, outerScopes = []) {
+  if (scope.referencesResolved) return scope;
+  const path = [scope].concat(outerScopes);
+  scope.refs.forEach(ref => {
```

Comment:
It might be worth measuring if a simple cache object for the lookup helps the performance at this point. I assume that in the majority of cases we have several refs of the same name inside the scope.